### ISSUE TITLE
Populate K12 school database and add filters

### ISSUE THREAD

**Issue by milescalabresi** [learning-unlimited/ESP-Website#3508]
We should grab school names from public sources and in order to make the available list more manageable have some kind of filter (e.g. the program sets a list of top cities/states/counties/zips or have students filter by state/city). Bonus points if the scraper updates automatically from time to time

**Comment by willgearty** [comment]
Looks like these are the resources we want:
Public school data: https://nces.ed.gov/ccd/files.asp (e.g., https://nces.ed.gov/ccd/Data/zip/ccd_sch_029_2021_w_1a_080621.zip) 
Private school data: https://nces.ed.gov/surveys/pss/pssdata.asp (e.g., https://nces.ed.gov/surveys/pss/zip/pss1920_pu_csv.zip)

Both should be fairly easy to filter and incorporate into the codebase. The private school data files appear to be easier to scrape in the future, since the file names are predictable (whereas the public school files have unique IDs).

Finally, there are >50,000 public and private middle and high schools across the country, so we would probably want to do the filtering server-side, perhaps like how we do the student/teacher dropdown searchboxes?

**Comment by milescalabresi** [comment]
Thanks for finding those. Messy data scraping is something I know how to do :) Good point about doing it server-side

**Comment by hwatheod** [comment]
For a large list like that, instead of a dropdown, can we do auto-complete instead? Even auto-complete should be guided by some location information like zip code or at least state.

**Comment by willgearty** [comment]
Yeah, that's exactly what I was thinking @hwatheod, just like how we do the student and teacher auto-completes.

**Comment by SuyashJain17** [comment]
Hi @hwatheod @milescalabresi @willgearty — I’ve been reading through this issue and I agree that a dropdown with 50k+ schools would probably be overwhelming (and slow).

It seems like switching to a server-side autocomplete — similar to the student/teacher search — would make a lot more sense. Maybe we could also require a state or a few characters of the school name before triggering the search, just to keep queries efficient.

For the data itself, we could start by importing the NCES public + private school datasets into the existing K12School model, and then worry about automating updates later as a second step.

Does that sound like a reasonable direction? I’d be happy to draft a more concrete plan if so.

**Comment by hwatheod** [comment]
I've assigned you @SuyashJain17 .

One concern I have is, are we going to put 50K `K12School` entries on every program's server? It seems wasteful, especially since programs generally only work with their local schools. But it is not totally clear how to limit it. A program located in a small state may attract students from neighboring states, and large programs may occasionally even attract students from very far away. 

Maybe I am overestimating the costs here.

**Comment by SuyashJain17** [comment]
That’s a fair concern. My understanding is that ~50k rows is still fairly small for a database table, especially with proper indexing, so the storage and performance impact should be minimal.

I think importing the full dataset might be the simplest starting point, but I’m happy to explore limiting it by state if that feels safer.

**Comment by hwatheod** [comment]
I think we can take up the dataset decision later. Does it affect your design?
